packages/next/build/webpack/plugins/telemetry-plugin.ts

```diff
@@ -0,0 +1,111 @@
+import type { Compiler, Compilation } from 'webpack'
+
+type Feature = 'next/image' | 'next/script' | 'next/dynamic'
+
+interface FeatureUsage {
+  featureName: Feature
+  invocationCount: number
+}
+
+/**
+ * A vertex in the module graph.
+ */
+interface Module {
+  type: string
+  identifier(): string
+}
+
+/**
+ * An edge in the module graph.
+ */
+interface Connection {
+  originModule: unknown
+}
+
+// Map of a feature module to the file it belongs in the next package.
+const FEATURE_MODULE_MAP: ReadonlyMap<Feature, string> = new Map([
+  ['next/image', '/next/image.js'],
+  ['next/script', '/next/script.js'],
+  ['next/dynamic', '/next/dynamic.js'],
+])
+
+/**
+ * Plugin that queries the ModuleGraph to look for modules that correspond to
+ * certain features (e.g. next/image and next/script) and record how many times
+ * they are imported.
+ */
+export class TelemetryPlugin {
+  private usageTracker = new Map<Feature, FeatureUsage>()
+
+  constructor() {
+    for (const featureName of FEATURE_MODULE_MAP.keys()) {
+      this.usageTracker.set(featureName, {
+        featureName,
+        invocationCount: 0,
+      })
+    }
+  }
+
+  apply(compiler: Compiler): void {
+    compiler.hooks.make.tapAsync(
```

Comment:
This hook is only available with webpack 5 if I recall correctly, do we need to only enable this plugin when webpack 5 is being used?

Reply:
`make` is available in webpack 4 according to the [docs](https://v4.webpack.js.org/api/compiler-hooks/#make).
I also looked at an existing plugin `FontStylesheetGatheringPlugin` that uses the same hook
https://github.com/vercel/next.js/blob/9d281421397d1de8e13d4c68c86ac83b641c27dd/packages/next/build/webpack/plugins/font-stylesheet-gathering-plugin.ts#L160
and we did not check `isWebpack5` in the config when initializing the plugin:
https://github.com/vercel/next.js/blob/9d281421397d1de8e13d4c68c86ac83b641c27dd/packages/next/build/webpack-config.ts#L1364-L1375

Reply:
Ah right, was thinking that might be why webpack 4 was failing, re-ran it locally and it seems this is the actual error:

```sh
~/dev/vercel/idk(telemetry-plugin) » NEXT_PRIVATE_TEST_WEBPACK4_MODE=1 yarn next build test/integration/production
yarn run v1.22.11
$ node --trace-deprecation --enable-source-maps packages/next/dist/bin/next build test/integration/production
info  - Using webpack 4 in Next.js is deprecated. Please upgrade to using webpack 5: https://nextjs.org/docs/messages/webpack5
warn  - You have enabled experimental feature(s).
warn  - Experimental features are not covered by semver, and may cause unexpected or broken application behavior. Use them at your own risk.

info  - Checking validity of types  
(node:82609) UnhandledPromiseRejectionWarning: /Users/jj/dev/vercel/idk/packages/next/build/webpack/plugins/telemetry-plugin.ts:63
              ).moduleGraph.getIncomingConnections(module)
                            ^

TypeError: Cannot read property 'getIncomingConnections' of undefined
    at null.<anonymous> (/Users/jj/dev/vercel/idk/packages/next/build/webpack/plugins/telemetry-plugin.ts:63:29)
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/Users/jj/dev/vercel/idk/packages/next/dist/compiled/webpack/bundle4.js:32108:10), <anonymous>:16:1)
    at AsyncSeriesHook.lazyCompileHook (/Users/jj/dev/vercel/idk/packages/next/dist/compiled/webpack/bundle4.js:32045:20)
    at Compilation.finish (/Users/jj/dev/vercel/idk/packages/next/dist/compiled/webpack/bundle4.js:39402:28)
    at /Users/jj/dev/vercel/idk/packages/next/dist/compiled/webpack/bundle4.js:41156:17
    at _done (eval at create (/Users/jj/dev/vercel/idk/packages/next/dist/compiled/webpack/bundle4.js:32108:10), <anonymous>:9:1)
    at eval (eval at create (/Users/jj/dev/vercel/idk/packages/next/dist/compiled/webpack/bundle4.js:32108:10), <anonymous>:608:22)
    at /Users/jj/dev/vercel/idk/packages/next/dist/compiled/webpack/bundle4.js:39334:12
    at /Users/jj/dev/vercel/idk/packages/next/dist/compiled/webpack/bundle4.js:39246:9
    at processTicksAndRejections (internal/process/task_queues.js:77:11)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:82609) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:82609) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
    at emitDeprecationWarning (internal/process/promises.js:180:11)
    at processPromiseRejections (internal/process/promises.js:249:13)
    at processTicksAndRejections (internal/process/task_queues.js:96:32)
```

Reply:
I see! Thanks for much for pointing that out! I've added a check for `isWebpack5` when initializing the plugin in the config.